File: BinaryAdd.c
#include <stdio.h>

void BinaryAdd(int a[], int b[], int n,int c[]) {
		int promote = 0;
		int i;
		int result;
		
		for (i = n - 1; i >= 0; i--) {
			result = a[i] + b[i] + promote;
			if (result >= 2) {
				result = result - 2;
				promote = 1;
			}
			else
				promote = 0;
			c[i + 1] = result;
		}
		c[0] = promote;
}

int main() {
	int a[4] = {1,0,1,1};
	int b[4] = {1,1,1,1};
	int c[5];
	
	BinaryAdd(a, b, 4, c);
	printf("%d%d%d%d + %d%d%d%d = %d%d%d%d%d\n", a[0], a[1], a[2], a[3], b[0], b[1], b[2], b[3], c[0], c[1], c[2], c[3], c[4]);
	return 0;
}

File: binary_add.py
#!/usr/bin/env python
# encoding: utf-8


def binary_add(array1, array2):
    """
    Consider the problem of adding two n-bit binary integers, stored in two n-element arrays A and B,
    in big-endian order.
    The sum of the two integers should be stored in binary form in a (n + 1)-element array C.
    :param array1: n-element array A
    :param array2: n-element array B
    :return: (n + 1)-element array C
    """
    assert len(array1) == len(array2)
    n = len(array1)
    promote = 0
    array3 = [0] * (n + 1)
    for i in range(n - 1, -1, -1):
        result = array1[i] + array2[i] + promote
        promote = result // 2
        result = result % 2
        array3[i + 1] = result
    array3[0] = promote
    return array3
